tests/ast/decoder.py

```diff
@@ -828,16 +871,75 @@ def decode_expr(self, expr: proto.Expr) -> Any:
+                ascending = self.decode_expr(expr.sp_dataframe_sort.ascending)
+                if expr.sp_dataframe_sort.cols_variadic:
+                    return df.sort(*cols, ascending)
+                else:
+                    return df.sort(cols, ascending)
+
+            case "sp_relational_grouped_dataframe_agg":
+                grouped_df = self.decode_expr(
+                    expr.sp_relational_grouped_dataframe_agg.grouped_df
+                )
+                exprs = self.decode_col_exprs(
+                    expr.sp_relational_grouped_dataframe_agg.exprs.args
+                )
+                # Check whether exprs is a list before using the * operator since exprs can be a Column.
+                if expr.sp_relational_grouped_dataframe_agg.exprs.variadic is True:
+                    return grouped_df.agg(*exprs)
+                else:
+                    return grouped_df.agg(exprs)
+
+            case "sp_relational_grouped_dataframe_apply_in_pandas":
+                # TODO: SNOW-1830603 Flesh out this logic when implementing UDTFs. Need to create a dict to maintain
+                #       all functions registered (here, `func`). Implement `decode_callable_expr`.
+                # func = self.decode_callable_expr(expr.sp_relational_grouped_dataframe_apply_in_pandas.func)
+                # grouped_df = self.decode_expr(expr.sp_relational_grouped_dataframe_apply_in_pandas.grouped_df)
+                # kwargs = self.decode_dsl_map_expr(expr.sp_relational_grouped_dataframe_apply_in_pandas.kwargs)
+                # output_schema = self.decode_expr(expr.sp_relational_grouped_dataframe_apply_in_pandas.output_schema)
+                # return grouped_df.apply_in_pandas(func, output_schema, **kwargs)
+                pass
+
+            case "sp_relational_grouped_dataframe_builtin":
+                grouped_df = self.decode_expr(
+                    expr.sp_relational_grouped_dataframe_builtin.grouped_df
+                )
+                cols = self.decode_col_exprs(
+                    expr.sp_relational_grouped_dataframe_builtin.cols.args
+                )
+                agg_name = expr.sp_relational_grouped_dataframe_builtin.agg_name
+                if (
+                    expr.sp_relational_grouped_dataframe_builtin.cols.variadic
+                    and isinstance(agg_name, list)
+                ):
+                    return grouped_df.function(*agg_name)(*cols)
+                else:
+                    return grouped_df.function(agg_name)(*cols)
+
+            case "sp_relational_grouped_dataframe_ref":
+                return self.symbol_table[
+                    expr.sp_relational_grouped_dataframe_ref.id.bitfield1
+                ][1]
+
             case "sp_table":
                 assert expr.sp_table.HasField("name")
                 table_name = self.decode_table_name_expr(expr.sp_table.name)
                 return self.session.table(table_name)
 
+            case "udtf":
```

Comment:
I prefer handling the UDTF logic in a different PR since it would be adding a dict similar to the symbol table but for functions -- it has to be different from the symbol table since they both use a different counter to keep track of assign statements and the function ID on the client-side.